File: construct-mobile-advert/src/ios/ConstructAd.swift
import GoogleMobileAds
import AdSupport

let GMAD_TEST_APPLICATION_ID = 	"ca-app-pub-3940256099942544~1458002511"

@objc(ConstructAd)
class ConstructAd : CDVPlugin
{
	var bannerAD: BannerAdvert?
	var interstitialAD: InterstitialAdvert?
	var videoAD: VideoAdvert?
    var userConsent: UserConsent?
    
    override init() {
    }
    
    func GetDeviceID () -> String
    {
        if (ASIdentifierManager.shared().isAdvertisingTrackingEnabled)
        {
            return ASIdentifierManager.shared().advertisingIdentifier.uuidString
        }
        else
        {
            return ""
        }
    }
    
    func CreateGADRequest(debug: Bool) -> GADRequest
    {
        let request = GADRequest()
     
        if (debug)
        {
            request.testDevices = [ GetDeviceID() ]
        }
        
        if (userConsent?.status == PACConsentStatus.nonPersonalized)
        {
            let extras = GADExtras()
            extras.additionalParameters = ["npa": "1"]
            request.register(extras)
        }
        return request
    }

	@objc(CreateBannerAdvert:)
	func CreateBannerAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		let unitID = command.arguments[0] as? String ?? ""
		let adSize = command.arguments[1] as? String ?? ""
        let debug = command.arguments[2] as? String ?? ""

		if unitID == ""
		{
         promise.reject(msg: "Unit ID not specified")
			return
		}

		if adSize == ""
		{
			promise.reject(msg: "Ad size not specified")
			return
		}
        
        let request = CreateGADRequest(debug: debug == "true")
        
        self.bannerAD = BannerAdvert(request: request, viewController: viewController, prom: promise, id: unitID, adSize: adSize)
	}

	@objc(ShowBannerAdvert:)
	func ShowBannerAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		self.bannerAD?.show(view: viewController.view, prom: promise)
	}

	@objc(HideBannerAdvert:)
	func HideBannerAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		self.bannerAD?.hide(prom: promise)
	}

	@objc(CreateInterstitialAdvert:)
	func CreateInterstitialAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)
		let unitID = command.arguments[0] as? String ?? ""
        let debug = command.arguments[1] as? String ?? ""

		if unitID == ""
		{
			promise.reject(msg: "Unit ID not specified")
			return
		}
        
        let request = CreateGADRequest(debug: debug == "true")
        
        self.interstitialAD = InterstitialAdvert(request: request, prom: promise, id: unitID)
	}

	@objc(ShowInterstitialAdvert:)
	func ShowInterstitialAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		self.interstitialAD?.show(viewController: viewController, prom: promise)
	}

	@objc(CreateVideoAdvert:)
	func CreateVideoAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)
		let unitID = command.arguments[0] as? String ?? ""
        let debug = command.arguments[1] as? String ?? ""

		if unitID == ""
		{
			promise.reject(msg: "Unit ID not specified")
			return
		}
        
        let request = CreateGADRequest(debug: debug == "true")

        self.videoAD = VideoAdvert(request: request, prom: promise, id: unitID)
	}

	@objc(ShowVideoAdvert:)
	func ShowVideoAdvert(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		self.videoAD?.show(viewController: viewController, prom: promise)
	}

	@objc(SetUserPersonalisation:)
	func SetUserPersonalisation(command: CDVInvokedUrlCommand)
	{
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		var status = command.arguments[0] as? String ?? ""

		userConsent?.SetUserStatus(status: status, promise: promise)
	}

	@objc(Configure:)
	func Configure(command: CDVInvokedUrlCommand) {
		let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)

		var id = command.arguments[0] as? String ?? ""
        let pubID = command.arguments[1] as? String ?? ""
        let privacyURL = command.arguments[2] as? String ?? ""
		let displayFree = (command.arguments[3] as? String ?? "") == "true"
        let showConsent = command.arguments[4] as? String ?? ""
        let debug = (command.arguments[5] as? String ?? "") == "true"
        let debugLocation = command.arguments[6] as? String ?? ""
        
		if id == ""
		{
			promise.reject(msg: "Application ID not specified")
			return
		}
        
        if pubID == ""
        {
            promise.reject(msg: "Publisher ID not specified")
            return
        }
        
        if privacyURL == ""
        {
            promise.reject(msg: "Privacy URL not specified")
            return
        }
        
		if debug == true
		{
			id = GMAD_TEST_APPLICATION_ID
		}
        
        let deviceID = debug == true ? GetDeviceID() : ""
        
        GADMobileAds.configure(withApplicationID: id)
        
        userConsent = UserConsent()
        
        userConsent?.UpdateUserConsent(deviceID: deviceID, debugLocation: debugLocation, pubID: pubID, displayFreeOption: displayFree, privacyURLString: privacyURL, promise: promise, showConsent: showConsent, plugin: self)
	}
    
    @objc(RequestConsent:)
    func RequestConsent(command: CDVInvokedUrlCommand)
    {
        let promise = CommandPromise(id: command.callbackId, comDelegate: self.commandDelegate)
        self.userConsent?.ShowUserConsentForm(promise: promise, plugin: self)
    }
}

File: construct-mobile-advert/www/ConstructAd.js
var exec = require('cordova/exec');

var CLASS_NAME = "ConstructAd";

exports.CreateBannerAdvert = function CreateBannerAdvert (id, size, debug, fn)
{
	var METHOD_NAME = "CreateBannerAdvert";
	debug = debug ? "true" : "false";
	var args = [id, size, debug];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.ShowBannerAdvert = function ShowBannerAdvert (fn)
{
	var METHOD_NAME = "ShowBannerAdvert";
	var args = [];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.HideBannerAdvert = function HideBannerAdvert (fn)
{
	var METHOD_NAME = "HideBannerAdvert";
	var args = [];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.CreateInterstitialAdvert = function CreateInterstitialAdvert (id, debug, fn)
{
	var METHOD_NAME = "CreateInterstitialAdvert";
	debug = debug ? "true" : "false";
	var args = [id, debug];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.ShowInterstitialAdvert = function ShowInterstitialAdvert (fn)
{
	var METHOD_NAME = "ShowInterstitialAdvert";
	var args = [];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.CreateVideoAdvert = function CreateVideoAdvert (id, debug, fn)
{
	var METHOD_NAME = "CreateVideoAdvert";
	debug = debug ? "true" : "false";
	var args = [id, debug];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.ShowVideoAdvert = function ShowVideoAdvert (fn)
{
	var METHOD_NAME = "ShowVideoAdvert";
	var args = [];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.Configure = function Configure (id, pubID, privacyPolicy, showFree, showConsent, debug, debugLocation, fn)
{
	
	var METHOD_NAME = "Configure";

	showFree = showFree ? "true" : "false";
	debug = debug ? "true" : "false";

	var args = [id, pubID, privacyPolicy, showFree, showConsent, debug, debugLocation];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.SetUserPersonalisation = function SetUserPersonalisation (mode, fn)
{
	var METHOD_NAME = "SetUserPersonalisation";
	var args = [mode];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}

exports.RequestConsent = function RequestConsent (fn)
{
	var METHOD_NAME = "RequestConsent";
	var args = [];
	function resolve (str)
	{
		fn(null, str);
	}
	function reject (str)
	{
		fn(str, null);
	}
	exec(resolve, reject, CLASS_NAME, METHOD_NAME, args);
}